### PR TITLE
[wait-for] Reconnect on agent bouncing

### PR DESCRIPTION
**Pull request by SimonRichardson**
We need to ensure that we handle any issues when the agent bounces that
cause the API server to be pulled down and put back up. This is required
when we do a controller upgrade and that causes the agent to bounce and
then the API server.

If the API server doesn't come back up, then the only option is then to die.

## QA steps

We want to ensure that the query never completes, so we insert false to
force it to retry until the timeout occurs.

In one terminal:

```sh
$ juju bootstrap lxd test
$ juju wait-for model test -m controller --query="false" --debug
```

In another terminal

```sh
$ juju upgrade-controller --build-agent
```

Notice that the command retries to connect.

### CODE REVIEW

File: cmd/plugins/juju-wait-for/model.go [hunk 1]
```diff
@@ -89,16 +89,19 @@ func (c *modelCommand) Init(args []string) (err error) {
...
-		Client:  client,
-		Timeout: c.timeout,
+		ClientFn: c.newWatchAllAPIFunc,
+		Timeout:  c.timeout,
 	}
-	err = strategy.Run(c.name, c.query, c.waitFor)
+	strategy.Subscribe(func(event EventType) {
+		switch event {
+		case WatchAllStarted:
+			// When a watch has started, we should prime all the local caches,
```
Comment (manadart) [inline cmd/plugins/juju-wait-for/model.go:99]: Is this something that is intentionally left out of this patch?
This method does nothing...

**Review by manadart — COMMENTED** [review]
QA worked OK, and code looks good aside from one question.

File: cmd/plugins/juju-wait-for/model.go (hunk 1 above)
Comment (SimonRichardson) [reply cmd/plugins/juju-wait-for/model.go:99]: Yes, it's for this PR https://github.com/juju/juju/pull/12196 where I need to evict the cached entities.

**Comment by SimonRichardson** [comment]
$$merge$$

**Review by manadart — APPROVED** [review]

**Pull request merged**